clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/safety/TrafficGuard.java

```diff
@@ -0,0 +1,377 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.netflix.spinnaker.clouddriver.safety;
+
+import static java.lang.String.format;
+
+import com.google.common.collect.ImmutableMap;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.impl.Preconditions;
+import com.netflix.spinnaker.clouddriver.core.services.Front50Service;
+import com.netflix.spinnaker.clouddriver.exceptions.TrafficGuardException;
+import com.netflix.spinnaker.clouddriver.model.Cluster;
+import com.netflix.spinnaker.clouddriver.model.ClusterProvider;
+import com.netflix.spinnaker.clouddriver.model.HealthState;
+import com.netflix.spinnaker.clouddriver.model.ServerGroup;
+import com.netflix.spinnaker.clouddriver.names.NamerRegistry;
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService;
+import com.netflix.spinnaker.moniker.Moniker;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import retrofit.RetrofitError;
+
+@Component
+public class TrafficGuard {
+  private static final String MIN_CAPACITY_RATIO = "traffic-guards.min-capacity-ratio";
+  private final Logger log = LoggerFactory.getLogger(getClass());
+
+  private final List<ClusterProvider<?>> clusterProviders;
+  private final Front50Service front50Service;
+  private final Registry registry;
+  private final DynamicConfigService dynamicConfigService;
+
+  private final Id savesId;
+
+  @Autowired
+  public TrafficGuard(
+      List<ClusterProvider<?>> clusterProviders,
+      Optional<Front50Service> front50Service,
+      Registry registry,
+      DynamicConfigService dynamicConfigService) {
+    this.clusterProviders = clusterProviders;
+    this.front50Service = front50Service.orElse(null);
+    this.registry = registry;
+    this.dynamicConfigService = dynamicConfigService;
+    this.savesId = registry.createId("trafficGuard.saves");
+  }
+
+  public void verifyInstanceTermination(
+      String serverGroupName,
+      List<String> instanceIds,
+      String account,
+      String location,
+      String cloudProvider,
+      String operationDescriptor) {
+    // TODO(rz): I opted out of migrating this method because it isn't used in
+    //  my current refactors. This method uses clouddriver search endpoint,
+    //  which would be a much larger refactor to bring over in this commit. I
+    //  would like to postpone such a refactor until it's actually needed.
+    throw new UnsupportedOperationException(
+        "verifyInstanceTermination method has not been migrated from Orca yet");
+  }
+
+  public void verifyTrafficRemoval(
+      String serverGroupName,
+      String account,
+      String location,
```

Comment:
I was going to port Orca's `Location` class over to Clouddriver, but it looks like we can just use a String here - it'd basically just be `ServerGroup#getRegion()`.